AX: WebKit should match other browsers, and the main-thread text implementation, and not generate text positions for SVG graphics elements

twilco opened the pull request:
#### 1856c3a7da49090338b273e575e05092e58096d2
<pre>
AX: WebKit should match other browsers, and the main-thread text implementation, and not generate text positions for SVG graphics elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=295802">https://bugs.webkit.org/show_bug.cgi?id=295802</a>
<a href="https://rdar.apple.com/155637541">rdar://155637541</a>

Reviewed by Joshua Hoffman.

This discrepancy was discovered when investigating an infinite word-navigation loop with VoiceOver in which we would
constantly cycle between the last two words. This fixes that issue, but is likely not the root cause — invesigation
will continue beyond this commit. But this behavior change is right regardless, as it matches other browsers and the
WebKit main-thread text implementation.

Minimal reproduction:

```
&lt;div style=&quot;width: 580px&quot;&gt;
&lt;svg&gt;
    &lt;path d=&quot;M0 0&quot; fill=&quot;#252526&quot;/&gt;
    &lt;path d=&quot;M1000 1000&quot; fill=&quot;#252526&quot;/&gt;
    &lt;path d=&quot;M50 50&quot; fill=&quot;#252526&quot;/&gt;
    &lt;path d=&quot;M10 10&quot; fill=&quot;#252526&quot;/&gt;
    &lt;path d=&quot;M42 42&quot; fill=&quot;#252526&quot;/&gt;
    &lt;path d=&quot;M9999 9999&quot; fill=&quot;#252526&quot;/&gt;
&lt;/svg&gt;
&lt;p&gt;The quick fox&lt;/p&gt;
&lt;/div&gt;
```

* LayoutTests/accessibility/mac/index-for-text-markers-svg-graphics-elements-expected.txt: Added.
* LayoutTests/accessibility/mac/index-for-text-markers-svg-graphics-elements.html: Added.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):

Canonical link: <a href="https://commits.webkit.org/297307@main">https://commits.webkit.org/297307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eef6a01a796b132fbe38db37b700dd6b197469b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111278 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21409 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117310 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61546 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113240 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84579 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114225 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25253 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100188 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65025 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24595 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18328 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61129 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94633 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18396 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120441 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28479 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93505 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38703 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96463 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93329 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23785 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38427 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16203 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34319 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38216 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43693 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37881 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41214 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39583 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1455,6 +1456,12 @@ AXTextRuns AccessibilityRenderObject::textRuns()
     }
 
     if (isReplacedElement()) {
+        if (is<SVGPathElement>(element())) {
```
This feels too narrow to me. I'd be really surprised if <path> has this bug but the other SVG elements that render some sort of shape on the screen do not, e.g.: `<rect>, <circle>, <ellipse>, <line>, <polyline>, <polygon>`.

Less obvious would be `<text>`, `<image>`, and `<use>`. Of these the only one I'd want to worry about not breaking is `<text>`, so my thought would be that as long as <text> doesn't break, I'd err on the side of excluding more.

What about checking for any subclass of SVGGraphicsElement here?

twilco replied:
Agree! Updated the testcase to include several subclass elements of `SVGGraphicsElement`, and added a `<text>` element and ensured it remains unbroken by this change.

hoffmanjoshua commented on LayoutTests/accessibility/mac/index-for-text-markers-with-svg-path-expected.txt:
```diff
@@ -0,0 +1,22 @@
+Text marker 11 had index 11
+Text marker 12 had index 12
+Text marker 13 had index 13
+Text marker 14 had index 0
```
Is checking that this is 0 the way we validate the loop? Or when we loop infinitely, are the above indices wrong too?

twilco replied:
There should be only 13 indices in this test case. The existing behavior of the AXIndexForTextMarker API is to return 0 for invalid markers. So checking zero here verifies there are no more indices or positions to be found.

Would you prefer we do this some other way?

twilco committed:
AX: WebKit should match other browsers, and the main-thread text implementation, and not generate text positions for SVG graphics elements
https://bugs.webkit.org/show_bug.cgi?id=295802
rdar://155637541

Reviewed by Joshua Hoffman.

This discrepancy was discovered when investigating an infinite word-navigation loop with VoiceOver in which we would
constantly cycle between the last two words. This fixes that issue, but is likely not the root cause — invesigation
will continue beyond this commit. But this behavior change is right regardless, as it matches other browsers and the
WebKit main-thread text implementation.

Minimal reproduction:

```
<div style="width: 580px">
<svg>
    <path d="M0 0" fill="#252526"/>
    <path d="M1000 1000" fill="#252526"/>
    <path d="M50 50" fill="#252526"/>
    <path d="M10 10" fill="#252526"/>
    <path d="M42 42" fill="#252526"/>
    <path d="M9999 9999" fill="#252526"/>
</svg>
<p>The quick fox</p>
</div>
```

* LayoutTests/accessibility/mac/index-for-text-markers-svg-graphics-elements-expected.txt: Added.
* LayoutTests/accessibility/mac/index-for-text-markers-svg-graphics-elements.html: Added.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):

Canonical link: https://commits.webkit.org/297307@main
